src/utils/object.js

```diff
@@ -44,3 +46,46 @@ export function map(opt_initial) {
 export function hasOwn(obj, key) {
   return hasOwn_.call(obj, key);
 }
+
+/**
+ * @param {!Object} a the destination object
+ * @param {!Object} b
+ * @param {number} currentDepth
+ * @param {number|undefined=} opt_maxDepth The maximum depth for deep merge, beyond which
+ *                               Object.assign will be used.
+ * @return {!Object}
+ */
+function deepMerge_(a, b, currentDepth, opt_maxDepth) {
+  if (currentDepth > opt_maxDepth) {
+    Object.assign(a, b);
+    return a;
+  }
+  Object.keys(b).forEach(key => {
+    const newValue = b[key];
+    // Perform a deep merge IFF both a and b have the same property and
+    // the properties on both a and b are non-null plain objects
+    if (hasOwn(a, key)) {
+      const oldValue = a[key];
+      if (isObject(newValue) && isObject(oldValue)) {
+        a[key] = deepMerge_(oldValue, newValue, currentDepth + 1, opt_maxDepth);
+        return;
+      }
+    }
+    a[key] = newValue;
+  });
+  return a;
+}
+
+/**
+ * Deep merge object b into object a. Both a and b can only contain
+ * primitives, arrays, and plain objects. For any conflicts, object b wins.
```

Comment:
We don't appear to enforce this (only primitives, arrays and plain objects). Remove?